File: frontend/src/types/network.ts
export interface Node {
  id: string;
  label: string;
  attributes: Record<string, any>;
}

export interface Edge {
  source: string;
  target: string;
  weight?: number;
  attributes?: Record<string, any>;
}

export interface NetworkData {
  network_id: number;
  name: string;
  nodes: Node[];
  edges: Edge[];
  directed: boolean;
  weighted: boolean;
}

export interface NetworkModel {
  id: number;
  name: string;
  description?: string;
  dataset_id: number;
  created_at: string;
  updated_at: string;
  node_count: number;
  edge_count: number;
  directed: boolean;
  weighted: boolean;
  attributes?: Record<string, any>;
}

export interface NetworkMetrics {
  network_id: number;
  global_metrics: {
    density: number;
    average_path_length: number;
    diameter: number;
    clustering: number;
    modularity?: number;
  };
  node_metrics: Record<string, {
    degree: number;
    betweenness?: number;
    closeness?: number;
    eigenvector?: number;
    clustering?: number;
  }>;
}

export interface Community {
  size: number;
  density: number;
  cohesion: number;
  nodes: string[];
}

export interface Communities {
  network_id: number;
  algorithm: string;
  num_communities: number;
  modularity: number;
  communities: Record<string, Community>;
}

export interface LayoutOptions {
  type: 'force' | 'circular' | 'hierarchical' | 'radial';
  parameters?: {
    [key: string]: any;
  };
}

export interface VisualizationOptions {
  layout: LayoutOptions;
  node_size: {
    by: 'uniform' | 'degree' | 'betweenness' | 'closeness' | 'eigenvector' | 'clustering' | 'attribute';
    attribute?: string;
    scale?: [number, number];
  };
  node_color: {
    by: 'community' | 'department' | 'role' | 'attribute' | 'centrality';
    attribute?: string;
    scale?: string[];
  };
  edge_width: {
    by: 'uniform' | 'weight';
    scale?: [number, number];
  };
  show_labels: boolean;
  label_property?: string;
  filters?: {
    departments?: string[];
    roles?: string[];
    min_tie_strength?: number;
  };
}

File: backend/app/schemas/network.py
from typing import Dict, List, Optional, Any
from datetime import datetime
from pydantic import BaseModel, Field


class Node(BaseModel):
    """Schema for network nodes."""
    id: str
    label: str
    attributes: Dict[str, Any] = Field(default_factory=dict)


class Edge(BaseModel):
    """Schema for network edges."""
    source: str
    target: str
    weight: Optional[float] = None
    attributes: Optional[Dict[str, Any]] = None


class NetworkBase(BaseModel):
    """Base network schema with common attributes."""
    name: str
    directed: bool = False
    weighted: bool = False
    description: Optional[str] = None


class NetworkCreate(NetworkBase):
    """Schema for network creation."""
    dataset_id: Optional[int] = None
    project_id: Optional[int] = None
    source_col: Optional[str] = None
    target_col: Optional[str] = None
    weight_col: Optional[str] = None


class NetworkUpdate(BaseModel):
    """Schema for network updates - all fields optional."""
    name: Optional[str] = None
    description: Optional[str] = None
    directed: Optional[bool] = None
    weighted: Optional[bool] = None
    attributes: Optional[Dict[str, Any]] = None


class NetworkInDB(NetworkBase):
    """Network schema as stored in database."""
    id: int
    created_at: datetime
    updated_at: datetime
    node_count: Optional[int] = None
    edge_count: Optional[int] = None
    dataset_id: Optional[int] = None
    project_id: Optional[int] = None
    file_path: Optional[str] = None
    attributes: Optional[Dict[str, Any]] = None
    metrics: Optional[Dict[str, Any]] = None
    
    class Config:
        orm_mode = True


class Network(NetworkInDB):
    """Full network schema for API responses."""
    pass


class NetworkData(BaseModel):
    """Schema for network visualization data."""
    network_id: int
    name: str
    nodes: List[Node]
    edges: List[Edge]
    directed: bool
    weighted: bool


class NetworkMetricsGlobal(BaseModel):
    """Schema for global network metrics."""
    density: float
    average_path_length: Optional[float] = None
    diameter: Optional[float] = None
    clustering: float
    modularity: Optional[float] = None


class NodeMetrics(BaseModel):
    """Schema for node-level metrics."""
    degree: float
    betweenness: Optional[float] = None
    closeness: Optional[float] = None
    eigenvector: Optional[float] = None
    clustering: Optional[float] = None


class NetworkMetrics(BaseModel):
    """Schema for network metrics."""
    network_id: int
    global_metrics: NetworkMetricsGlobal
    node_metrics: Dict[str, NodeMetrics]


class Community(BaseModel):
    """Schema for network communities."""
    size: int
    density: float
    cohesion: float
    nodes: List[str]


class Communities(BaseModel):
    """Schema for community detection results."""
    network_id: int
    algorithm: str
    num_communities: int
    modularity: float
    communities: Dict[str, Community]


class LayoutOptions(BaseModel):
    """Schema for network layout options."""
    type: str
    parameters: Optional[Dict[str, Any]] = None


class VisualizationOptions(BaseModel):
    """Schema for network visualization options."""
    layout: LayoutOptions
    node_size: Dict[str, Any]
    node_color: Dict[str, Any]
    edge_width: Dict[str, Any]
    show_labels: bool
    label_property: Optional[str] = None
    filters: Optional[Dict[str, Any]] = None